tests/lib/defer.sh

```diff
@@ -0,0 +1,73 @@
+#!/bin/sh
+
+# defer enqueues execution of shell commands.
+#
+# synopsis: defer [--scope=SCOPE] <COMMANDS>
+#
+# Given commands are stored in a shell script called .$SCOPE_defer_$N where
+# SCOPE is a name of the group of defer statements and N is an incrementing
+# counter. The counter is stored in the file .$SCOPE_defer_count, protected
+# from concurrent access by a flock-based lock file .defer_lock.
+#
+# Scope defaults to "auto", it exists to allow the prepare and restore code
+# to use deferrals without interfering with the use in the execute section.
+#
+# For the prepare/restore sections please use --scope=prepare. For the execute
+# section please use no scope at all. It is customary to use "trap run_deferred
+# EXIT" inside the execute section, to allow deferrals to clean up background
+# processes that would otherwise interfere with spread.
+#
+# To run all deferred commands use run_deferred with the same scope argument.
+defer() {
+    scope=auto
+    case "${1:-}" in
```

Comment:
any reason you're not using `getopt` instead?

Reply:
No strong reason, I just personally don't use it much. I'm happy to adjust this.